packages/@aws-cdk/assets/lib/fs/utils.ts

```diff
@@ -1,34 +1,113 @@
 import fs = require('fs');
 import minimatch = require('minimatch');
 import path = require('path');
+import { CopyOptions } from './copy-options';
 import { FollowMode } from './follow-mode';
 
 /**
  * Determines whether a given file should be excluded or not based on given
  * exclusion glob patterns.
  *
  * @param exclude  exclusion patterns
- * @param filePath file apth to be assessed against the pattern
+ * @param filePath file path to be assessed against the pattern
  *
  * @returns `true` if the file should be excluded
  */
 export function shouldExclude(exclude: string[], filePath: string): boolean {
-  let excludeOutput = false;
+  const [_shouldExclude] = shouldExcludePriority(exclude, filePath);
+  return _shouldExclude;
+}
 
-  for (const pattern of exclude) {
+/**
+ * Determines whether a given file should be excluded or not based on given
+ * exclusion glob patterns.
+ *
+ * @param exclude  exclusion patterns
+ * @param filePath file patg to be assessed against the pattern
+ *
+ * @returns `true` if the file should be excluded, followed by the index of the rule applied
+ */
+export function shouldExcludePriority(exclude: string[], filePath: string): [boolean, number] {
+  return exclude.reduce<[boolean, number]>((res, pattern, patternIndex) => {
     const negate = pattern.startsWith('!');
     const match = minimatch(filePath, pattern, { matchBase: true, flipNegate: true });
 
     if (!negate && match) {
-      excludeOutput = true;
+      res = [true, patternIndex];
     }
 
     if (negate && match) {
-      excludeOutput = false;
+      res = [false, patternIndex];
     }
-  }
 
-  return excludeOutput;
+    return res;
+  }, [false, -1]);
+}
+
+/**
+ * Determines whether a given file should be excluded,taking into account deep file structures
+ *
+ * @param exclude  exclusion patterns
+ * @param filePath file path to be assessed against the pattern
+ */
+export function shouldExcludeDeep(exclude: string[], relativePath: string): boolean {
+  const [_shouldExclude] = relativePath.split(path.sep).reduce<[boolean, number, string]>(
+    ([accExclude, accPriority, pathIterator], pathComponent) => {
+      pathIterator = path.join(pathIterator, pathComponent);
```

Comment:
I feel it would be cleaner if you made a function like `pathComponents()` (which given a path like `a/b/c` would return `[a, a/b, a/b/c]`) and iterated over that.

Again I feel a simple `for` loop would be more natural. If you really want to make it functional, why not something like:

```ts
const resultsAndPrios = pathComponents(relativePath).map(pc => shouldExcludePriority(rules,  pc));
const [shouldExclude] = pickHighest(resultsAndPrios, x => x[1]);
return shouldExclude;
```

Requires you to write some helpers, but to me is a lot clearer what is happening.

Reply:
I did some refactoring, preferring `for` loops over `reduce`s. I've only kept `map`s when I needed to know both the current array value and index.